indexing-service/src/test/java/io/druid/indexing/overlord/http/OverlordResourceTest.java

```diff
@@ -0,0 +1,305 @@
+        taskLockbox,
+        taskStorage,
+        taskActionClientFactory,
+        druidNode,
+        indexerZkConfig,
+        new TaskRunnerFactory()
+        {
+          @Override
+          public TaskRunner build()
+          {
+            return new MockTaskRunner(taskCountDownLatches);
+          }
+        },
+        new ResourceManagementSchedulerFactory()
+        {
+          @Override
+          public ResourceManagementScheduler build(
+              RemoteTaskRunner runner, ScheduledExecutorFactory executorFactory
+          )
+          {
+            return new NoopResourceManagementScheduler();
+          }
+        },
+        curator,
+        new NoopServiceAnnouncer(){
+          @Override
+          public void announce(DruidNode node)
+          {
+            announcementLatch.countDown();
+          }
+        },
+        serviceEmitter
+    );
+    EmittingLogger.registerEmitter(serviceEmitter);
+  }
+
+  @Test
+  public void testOverlordResource() throws Exception{
+    // basic task master lifecycle test
+    taskMaster.start();
+    Assert.assertTrue(announcementLatch.await(2, TimeUnit.SECONDS));
+    // To be on safer side wait for "leading" boolean variable to be set to true
+    Thread.sleep(10);
+    Assert.assertTrue(taskMaster.isLeading());
+    Assert.assertEquals(taskMaster.getLeader(), druidNode.getHostAndPort());
+    // Test Overlord resource stuff
+    overlordResource = new OverlordResource(taskMaster, new TaskStorageQueryAdapter(taskStorage), null, null, null);
+    Response response = overlordResource.getLeader();
+    Assert.assertEquals(druidNode.getHostAndPort(), response.getEntity());
+    String taskId_0 = "0";
+    NoopTask task_0 = new NoopTask(taskId_0, 0, 0, null, null);
+    response = overlordResource.taskPost(task_0);
+    Assert.assertEquals(200, response.getStatus());
+    Assert.assertEquals(ImmutableMap.of("task", taskId_0), response.getEntity());
+    // Duplicate task - should fail
+    response = overlordResource.taskPost(task_0);
+    Assert.assertEquals(400, response.getStatus());
+    // Task payload for task_0 should be present in taskStorage
+    response = overlordResource.getTaskPayload(taskId_0);
+    Assert.assertEquals(task_0, ((Map) response.getEntity()).get("payload"));
+    // Task not present in taskStorage - should fail
+    response = overlordResource.getTaskPayload("whatever");
+    Assert.assertEquals(404, response.getStatus());
+    // Task status of the submitted task should be running
+    response = overlordResource.getTaskStatus(taskId_0);
+    Assert.assertEquals(taskId_0, ((Map) response.getEntity()).get("task"));
+    Assert.assertEquals(TaskStatus.running(taskId_0).getStatusCode(), ((TaskStatus)((Map) response.getEntity()).get("status")).getStatusCode());
+    // Simulate completion of task_0
+    taskCountDownLatches[Integer.parseInt(taskId_0)].countDown();
+    // Wait for taskStorage sync
+    Thread.sleep(150);
```

Comment:
I think countdown latch cannot be used here as TaskQueue schedules a periodic thread which syncs from storage, there's no condition to wait on.